### ISSUE TITLE
nothing known about [module]

### ISSUE THREAD

**Issue by Satyam** [yui/yogi#53]
I've got `nothing known about`  from a non-gallery module that is under a repository cloned from GitHub, This repository is not related to the Gallery, but it is a fork of someone else's.  The folders are structured just as they would be in Gallery modules and it compiles fine using yogi and the result is a well-formed YUI module with its YUI.add() wrapper and the dependencies listed fine.  In fact, I can use that module fine.

However, when I try to run any tests, it doesn't know where to find the built files for the module being tested.  I can include that with a `<script>` tag and it runs fine, but then I cannot run the coverage tests without changing that `<script>` to use the `-coverage` version.  Anyway, I don't need to do this for a regular Gallery module, I just tried it because nothing else worked.

I tried a `yogi info` and this is what it says:

```
cd /home/satyam/git/jivaprototype/src/zeView/
$ /usr/bin/yogi info
yogi [info]  using yogi@0.0.66 on node@0.8.18
yogi [info]  using module: zeView
yogi [info]  showing info for zeView
yogi [bail] nothing known about zeView
yogi exited, code: 1
```

After that I did a `yogi init` to add a new empty module and it creates everything fine. Comparing the files for my actual module and the fake one I created just for testing, I have all the equivalent files, build.json, the meta file, unit testing and so on.

In the info `yogi init` provides prior to doing the creation it shows everything right, the GitHub repository, the local path, everything as `yogi info` should.   But then, if i actually do a `yogi info` the same thing happens, it knows nothing about that module.  If I add the fake, empty module as a dependency of the tests `yogi init` created for the module, it doesn't find it either.

With `yogi init` failing, I cannot do --set-repo or --set-path and, anyway, when doing `yogi init` to create the new empty module, it showed everything fine so I shouldn't have needed to change those since `yogi init` knew about them.

I must have something missing and I don't know what.

Thanks in advance for your help.

**Comment by davglass** [comment]
I'm not following here, it's a "non-gallery" module and you expect
gallery commands like "yogi info" to work?

**Comment by Satyam** [comment]
Nope, I want `yogi test` to work, just did the `yogi info` to gather information that might be useful to solve this issue.   Since yogi wasn't able to locate the module being tested, I wondered what was it that yogi knew about the module and `yogi info` said it didn't know anything.   I also tried `yogi loader`  assuming that the information that it could gather might help `yogi test` to locate the build files for the module being tested.

**Comment by davglass** [comment]
Can you please provide step by step instructions to replicate.

**Comment by Satyam** [comment]
In any repo not already associated with Gallery or YUI, do a `yogi init qq`.   The command will ask to confirm the configuration for the new module.  Move to that module and start `yogi serve`.  
Edit the `qq.js` file, add something easy to check like `Y.QQ = 'hello'`.  
Change the /tests/unit/index.html file so that it adds a dependency for module qq.  Same in the requires of the `module-test.js` file.  Change the test to something like `Y.Assert.areEqual('hello', Y.QQ);`
Compile the module and run the test.  yogi doesn't even try to load `qq.js`.

The console will show `yui: NOT loaded: qq` and the Net tab will show no attempt to load qq.js.

Adding this to the `index.html` makes it work, however, that is not required of Gallery modules:

```
groups: {
        qq: {
            base: ' ../../../build/',
            modules: {
                qq: {
                    path: 'qq/qq-min.js'
                }
            }
        }
    }
```

With this added, both the regular test and the coverage work (all run from `yogi serve`), but grover doesn't (not an issue in my case, whatever tests it can run is fine with me) and `module info` says it knows nothing about that module, which is also fine.  However, the last one is funny since `yogi init` was quite able to find out the info it needed.

`yogi serve` does something funny about the paths to the files since once I added the groups and modules shown above,  it reports the path to qq.js as:  `http://localhost:5000/build/qq/qq.js` so, I am really not sure who is parsing that path in the YUIconfig since the server is being started at `src/qq` .

I don't mean to get all the functionality of yogi, I've made some of the tests above just to probe what works and what doesn't and provide more info.  I just want to run shifter, which I can, run tests and coverage.  I don't mind not having grover or module info.  And I would rather have to make as little changes as possible if, at any time, the module is worth putting in the Gallery.

**Comment by davglass** [comment]
Alright, I think I have it now.

This is not really an issue with yogi, more of an information issue &
something lacking in Loader.

For Gallery modules, yogi "bends" the rules since any module with a
gallery- prefix is covered by a default pattern in Loader. That means
Loader will always attempt to fetch the module even without meta-data.
Then yogi does her "magic".

Since there is no pattern & no meta-data with this module, Loader will
fail with the "not loaded" message.

Make sense?

**Comment by Satyam** [comment]
Could yogi bend the rules for non-Gallery modules, perhaps after a `yogi loader` is done?  All the pieces seem to be there, different parts of yogi do manage to find things out.  `yogi init` shows plenty of module info, `yogi loader` can read the modules and their dependencies, shifter has no problem compiling everything.  The only thing it would not be able to do is to load modules from one repo into code running in another.  Perhaps a list of repos to check for modules might be added to the `.yogi` configuration file.

Anyway, just an idea, and a low priority one at that.  Thanks.

**Comment by davglass** [comment]
I'm not sure that yogi can do that without hurting the process. Yogi assumes a lot of things, but she does them in a non-destructive manor. For non-core and non-gallery modules the best thing to do it statically include your module in the test file or create the YUI config to tell yogi about the module.

I "might" be able to change up `yogi init` to have it add the module config into the YUI constructor in the test file.

**Comment by Satyam** [comment]
My idea was to keep the `index.html` file just alike, whether a Gallery module or not.  

There is another use case that you might want to consider, and that is with a Gallery module.  Sometimes you are developing a couple of modules in parallel, a core and an a full version, for example, my `gallery-flyweight-tree` and `gallery-fwt-treeview` or Todd Smith's `datatable-editable` and the two actual cell editors, the inline and popup versions.  In the tests for the full versions you have to manually include the core part or it may either fail or, if it has ever been published, it will pick the version in the CDN, which might not be the one you are working on.  This cannot be solved by just adding the module definition of the module being tested itself since you need the module definition of the dependency as well.

I don't know what happens when such paired modules get uploaded and get deployed.   If both have been updated in the same cycle and the dependent gets tested before the dependency, the deployment process might give an error.  Anyway, this is another issue.

So, I thought that yogi might check the files produced by `yogi loader` in the repository and, perhaps, other repositories you add to the `.yogi` configuration file, and serve components from those.  More or less as it does with the yui3 core files.  Just ideas, not really sure any of them are worth the pixels they are being displayed in.

**Comment by davglass** [comment]
Yeah, collaboration across multiple modules is a pain point. I've got some thoughts on it, but I've not closed in on them all yet. I'll leave this open as I think about it :)
